### DISCUSSION TITLE
Is it possible to upgrade to Keylcoak 26 and keycloak-js module, but at the same time serve /js/keycloak.js for older customers for a transitiont time?

### DISCUSSION THREAD

**Discussion by Nefcanto** [keycloak/keycloak#35989] [Q&A] [answered]
### Description

One step to upgrade Keycloak to version 26, is to change the way we use Keycloak JavaScript client.
In older versions, we had to include it from `https://accounts.example.com/js/keycloak.js`.
Now we should `npm install keycloak-js` then `import Keycloak from "keycloak-js"`.

The problem is that the upgrade seems to be very hard for a lot of clients who are connected to a single instance of Keycloak with lots of realms and clients.

Let's consider an instance of the Keycloak. It's in version 24. It server `/js/keycloak`. It has a thousand realms. Each realm has multiple clients some of which are React clients. They all have the `/js/keycloak` hardcoded. It's in their production code.

Now we want to upgrade that instance to 26. As soon as we do that, they all fail because `/js/keycloak.js` no longer exists.

And it's also impossible for us to upgrade all those 1000 deployed React apps simultaneously.
What should we do? Can we get `/js/keycloak.js` back for some major versions?

**Comment by jonkoops** [comment]
You could always build a version of Keycloak JS yourself (e.g. using [Rollup](https://rollupjs.org/)) and serve it through a proxy so it is available from the same endpoint, that should work just fine. Or even serve the old version by downloading it from the previous distribution, although that will of course not get any future fixes.

**Comment by Nefcanto** [comment]
@jonkoops, that would be awesome. Can you please show me how? Is there a doc for it? I have not done something similar before.

**Comment by Nefcanto** [comment, marked as answer]
Nevermind. I figured it out using AI's help. For anyone having this problem:

First, install the required packages at a temp directory:

```
npm install rollup rollup-plugin-terser @rollup/plugin-node-resolve @rollup/plugin-commonjs --save-dev
npm install keycloak-js
```

Then create a file called `rollup.config.js` and paste this code in it:

```
import { terser } from 'rollup-plugin-terser';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';

export default {
  input: 'node_modules/keycloak-js/lib/keycloak.js',
  output: {
    file: 'dist/keycloak.js',
    format: 'umd',
    name: 'Keycloak',
    sourcemap: true,
  },
  plugins: [
    resolve(),  // Resolves node_modules dependencies
    commonjs(), // Converts CommonJS to ES Modules
    terser()    // Minifies the output
  ]
};
```

Then add this snippet to your `package.json`:

```
"scripts": {
  "build": "rollup -c"
}
```

Then run:

```
npm run build
```

Now the `keycloak.js` is in the `dist` directory. Copy it wherever you want. I'm using Nginx. Thus I added this `location` before the reverse proxy. This way the `keycloak.js` would be served as a static file and the rest of the requests would be sent to the Keycloak itself:

```
    location /js/keycloak.js {
        root /some/path;
        index keycloak.js;
    }
```

**Reply by jonkoops** [reply]
FYI you don't need `@rollup/plugin-commonjs` as Keycloak JS is authored in spec compliant modules, not legacy CommonJS. Additionally, the `rollup-plugin-terser` can be replaced by the more modern `@rollup/plugin-terser`.

**Reply by Nefcanto** [reply]
@jonkoops, thank you so much. It's temporary. After upgrading the clients this would be removed.
